Felix/safety DRAFT

FK3wong opened the pull request:
First draft of Firmware to interface with the INA226 chip on the safety board.
Looking for feedback as I am not experienced with embedded programming.
Have not been able to test as the board was not ready when I was in Waterloo and is not currently a high priority.

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+
+int inRange(float upperLim, float lowerLim, float val)
+{
+    int inRange = (val > upperLim || val< lowerLim) ? 1: 0;
```
Logic is wrong, should be (val > upperLim || val< lowerLim) ? 0 : 1; or  (val < upperLim && val > lowerLim) ? 1 : 0

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+    i2c.read(SENSOR_SLAVE_ADDRESS, cmd, 2, false);
+    
+    char16_t currentData = cmd[1]; 
+    currentData = (cmd[2] << 8 | 0x00);
```
Are you trying to concatenate the two bytes here? The way this is written now only shifts the second byte and discards the first byte.

FelixWLWong replied:
yeah I think that was my intention. Can I instead do : `u_int16_t currentData = (cmd[2] << 8 | cmd[1]);
`

qinyang-bao replied:
yes that should work

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+    //may need char16_t?
+    char cmd[3] = {CURRENT_REGISTER};
+    
+    i2c.write(SENSOR_SLAVE_ADDRESS, cmd, 2);
```
when reading from this device, you write the register address, which is 1 byte  (not 2), and then read from it

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+float getCurrentData()
+{
+    //some conversion factor for current data
+    float conversionFactor = 1;
```
please specify the unit for conversion

FelixWLWong replied:
After reading the data sheet some more I have come across this section 

> For this example, the Shunt Voltage Register contains a value of 8,000 (representing 20 mV), which is multiplied by the Calibration Register value of 2560 and then divided by 2048 to yield a decimal value for the Current Register (04h) of 10000, or 2710h. Multiplying this value by 1 mA/bit results in the original 10-A level stated in the example.

Would this mean that I would not need a conversion factor as the data I read form this register would be in units of mA and any conversion I do would simply convert the value to Amps?

qinyang-bao replied:
This depends, shown on pg15 and 16 of data sheet, it appears that you need to determine the current LSB using the maximum expected current, and set the calibration register based on the LSB and Rshunt. Pls check with Ayesha about the maximum expected current and Rshunt.

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+            led2 = 1;
+        }
+
+        float measuredTemp = readThermosistor();
```
your readThermosistor function takes vin and vout as argument, you should supply them. Seems like you are using the  thermistor on the safety board, vin is 5V and vout is an analog input you can read at pin TEMP_ANLG_IN

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+const int SENSOR_SLAVE_ADDRESS = 0x1c << 1;
+
+enum configurationMode{
+    powerReset = 0,
```
as a convention, enums are in all caps to differentiate from variables

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+    int dataByte = 0x00;
+
+    //fill up register with configuration data
+    switch (operationMode)
```
this switch statement seems redundant, it's no different from dataByte |= powerReset

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
-const int k_interval_ms = 500;
+//below are settings for the INA226 chip. The default values have been selected 
+//but different conversion times can be selected as well. See datasheet for other settings
+const int busVoltageConversionTime = 4;
```
constants should be all caps as convention

FelixWLWong replied:
do I put underscores between words for constants?

qinyang-bao replied:
yes

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+const float rResistor = 0;
+
+//measurement thresholds, needs updating
+float currentUpperLim = 1;
```
thresholds should be constants as well

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+int configureSensor(configurationMode operationMode, int resetRegisters)
+{
+    char cmd[3] = {CONFIG_REGISTER, 0x00, 0x00};
+    int dataByte = 0x00;
```
use uint16_t instead

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+    }
+    
+    // give cmd the data to write 
+    cmd[1] |= dataByte;
```
there is no point in using |=, should just use =; also assigning an int that holds a value larger than what can be held in a char to a char results in undefined behavior. A better way would be cmd[1] = dataByte & 0xFF;

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+        if(inRange(tempUpperLim, tempLowerLim, measuredTemp))
+        {
+            led3 = 1;
+        }
```
If this is a testing app, better to print the values you read on serial as well

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+int calibrateSensor();
+int configureSensor(configurationMode operationMode, int resetRegisters);
+void initCAN();
+void processCANMsg(CANMsg *p_newMsg);
```
Remember to make this into a library after the changes

FelixWLWong replied:
By Library do you mean specifically a class or just a set or functions

qinyang-bao replied:
I would say class is better, look at MoistureSensor for reference

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+        u_int8_t command = 0;
+        if(can.read(rxMsg))
+        {
+            processCANMsg(&rxMsg);
```
This is not good naming, by the name of "processCANMsg", it suggests that the function would already extract the data in the received message.

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+        float measuredCurrent = getCurrentData();
+        if (inRange(currentUpperLim, currentLowerLim, measuredCurrent))
+        {
+            led1 = 1;
```
It's probably better to turn off the LEDs at some point in one loop, so you know if a read is successful each time

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
+    float rthermo = rResistor *((vin/vout)-1);
+
+    //convert resistance to temp (steinhart and hart eqn)
+    return 1/(A + B * log(rthermo) + C * pow(log(rthermo),3));
```
Data sheet for the thermistor we are using: 
https://industrial.panasonic.com/cdbs/www-data/pdf/AUA0000/AUA0000C8.pdf, 
also, research B value and how to use that to calculate temperature

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,219 @@
-const int k_interval_ms = 500;
+//below are settings for the INA226 chip. The default values have been selected 
+//but different conversion times can be selected as well. See datasheet for other settings
+const int busVoltageConversionTime = 4;
```
please add comment about what do the values for busVoltageConversionTime and shuntVoltageConversionTime  represent, saying they are default doesn't mean much

FelixWLWong commented:
Code is revised for a second round of critiques . Code is not complete as I am a bit unsure about how I should proceed with certain parts of the code such as creating data for each of the components the INA_226 chip will be sued in.

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,107 @@
+constexpr float TEMP_LOWER_LIM = 0;    
+
+//sample struct for component specific data
+struct elbow_data
```
call it elbow_config for consistency for this type of struct. Look at the actuator class for reference. It's prob also better to accept this struct in the constructor of the current sensor class

qinyang-bao commented on lib/sensors/INA_226.h:
```diff
@@ -0,0 +1,57 @@
+
+        //  I2C bus initialization
+        I2C i2c_;
+        //may need to adjust function to take sensor slave addresses as there many be many current sensors
```
typo, "may" be; also I do think this is the case

qinyang-bao commented on lib/sensors/INA_226.cpp:
```diff
@@ -0,0 +1,79 @@
+{
+    shunt_resistance_ = shunt_resistance;
+    max_expected_current_ = max_expected_current;
+    sensor_address_ = sensor_address << 1;
```
need to recalculate current lsb in the constructor

qinyang-bao replied:
this line should have a comment explaining why the address is shifted

qinyang-bao commented on lib/sensors/INA_226.h:
```diff
@@ -0,0 +1,57 @@
+
+        //current conversion constexprants
+        float max_expected_current_ = 0; 
+        float current_lsb = max_expected_current_ / pow(2,15); 
```
this will not work as expected as you will set max_expected_current_  in the constructor. Also usually you don't initialize the class members that will be set by the constructor. If you want default values, then use default argument or constructor overloading.

FK3wong replied:
What should I do for the Constants I have defined? Do I have them as a normal variable and then set them in them in the constructor?

qinyang-bao replied:
yes

qinyang-bao commented on lib/sensors/INA_226.cpp:
```diff
@@ -0,0 +1,79 @@
+{
+    float cal  = 0.00512/(current_lsb_ * shunt_resistance_); //calibration register, multiplied w/ current register value to get actual current in Amps
+
+    u_int8_t cmd[3] = {CALIBRATION_REGISTER};
```
cmd only needs to have one element in this case

FK3wong replied:
Why would it only need 1? Don't I have to write the address and 2 bytes of data to calibrate? Or am I misunderstanding how to use the write?

FK3wong replied:
Also what type is best for cmd[] ?

qinyang-bao replied:
sorry, i saw the line i2c_.write(sensor_address_, cmd, 1) and thought you were just writing one byte. Yes you would still need 3 bytes. But also, your cmd[2] value is undefined at this moment

qinyang-bao commented on lib/sensors/INA_226.cpp:
```diff
@@ -0,0 +1,79 @@
+float Safety_Board::getCurrentData()
+{
+    //may need char16_t?
+    u_int8_t cmd[3] = {CURRENT_REGISTER};
```
cmd does not need to have 3 element, 2 is enough since the read is going to overwrite the value that was set to write

qinyang-bao commented on lib/sensors/INA_226.cpp:
```diff
@@ -0,0 +1,79 @@
+    i2c_.read(sensor_address_, cmd, 2, false);
+    
+    u_int16_t voltage_data = (cmd[2] << 8 | cmd[1]);
+    
```
you do still need the conversion ratio. "The LSB for the Bus Voltage Register (02h) is a fixed 1.25 mV/bit", top of pg 16 on data sheet

qinyang-bao commented on lib/sensors/INA_226.cpp:
```diff
@@ -0,0 +1,79 @@
+    i2c_.read(sensor_address_, cmd, 2, false);
+    
+    u_int16_t current_data = (cmd[2] << 8 | cmd[1]);
+
```
bottom of pg 15 of data sheet, "The Current Register (04h) is then calculated by multiplying the decimal value of the Shunt Voltage Register (01h) contents by the decimal value of the Calibration Register and then dividing by 2048, as shown in Equation 3." And then this value should be multiplied to the current_LSB

FK3wong replied:
I'm a bit confused. Is it correct that the multiplication of the shunt voltage register by the Calibration register divided by 2048 is done and the value is but in the current register and what I need to do is multiply what is in the current register by the current_LSB?

qinyang-bao replied:
I think so

qinyang-bao commented on lib/sensors/thermistor.h:
```diff
@@ -0,0 +1,22 @@
+
+class thermistor{
+    private:
+        constexpr float B = 4700; //kelvins. Based on Resistances of 25 and 50 , part number ERT-J1VV104J
```
for all these members except ROOM_TEMP should be passed in from the constructor rather than hardcoded. This is a class for thermistor, not just ERT-J1VV104J. I would also recommend to rename the class sth like VoltageDividerThermistor or add comment that this is a class for thermistors that works by using a voltage divider

qinyang-bao commented on lib/sensors/thermistor.h:
```diff
@@ -0,0 +1,22 @@
+        constexpr float THERMISTOR_ROOM_TEMP_RESISTANCE = 100000; //ohms
+        constexpr float voltage_divider_resistance_ = 100000; //ohms
+    public:
+    AnalogIn thermistor_voltage_;
```
why is this member public

qinyang-bao commented on lib/sensors/thermistor.cpp:
```diff
@@ -0,0 +1,27 @@
+
+thermistor::thermistor(PinName thermistor_vout_pin): thermistor_voltage_(thermistor_vout_pin);
+
+float thermistor::readThermistor()
```
change functuon name to readThermistorDegrees

FK3wong replied:
Should I specify celcius/kelvin in the name as well?

qinyang-bao replied:
oh crap forgot both can be called degrees...maybe readThermistorCelcius then

ChristopherA057 commented on apps/safety/main.cpp:
```diff
@@ -1,13 +1,107 @@
+bool inRange(float upper_lim, float lower_lim, float val)
+{
+    bool in_range = (val > upper_lim || val< lower_lim) ? 0: 1;
+    return in_range;
```
Isn't it possible just to return (val > upper_lim || val < lower_lim) without having to create the in_range variable?

ChristopherA057 commented on lib/sensors/INA_226.cpp:
```diff
@@ -0,0 +1,79 @@
+#include "mbed.h"
```
mbed.h is included in INA_226.h already

ChristopherA057 commented on lib/sensors/INA_226.cpp:
```diff
@@ -0,0 +1,79 @@
+//may need to pass in shunt_resistance
+int Safety_Board::calibrateSensor()
+{
+    float cal  = 0.00512/(current_lsb_ * shunt_resistance_); //calibration register, multiplied w/ current register value to get actual current in Amps
```
I might be missing something obvious, but shouldn't .00512 be replaced by a const so that the code is more self-explanatory?

ChristopherA057 commented on lib/sensors/thermistor.cpp:
```diff
@@ -0,0 +1,27 @@
+#include "mbed.h"
```
mbed.h is included in themistor.h already

a-gratton commented on lib/sensors/thermistor.cpp:
```diff
@@ -0,0 +1,27 @@
+    */ 
+   float measured_thermistor_temp_k = (ROOM_TEMP * B) / (B + (ROOM_TEMP * log((thermisterResistance / THERMISTOR_ROOM_TEMP_RESISTANCE))));
+
+   return measured_thermistor_temp_k - 273.15;
```
Might be better to define a constant ABSOLUTE_ZERO

FK3wong commented:
Added some more fixes.
Added some more functionality to use some of the other functions of the INA_226 chip.
Looking for some comments on how I can make the app more user friendly

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -40,8 +34,27 @@ bool inRange(float upper_lim, float lower_lim, float val);
+    ComponentConfig turntable_config = {CUR_SEN_I2C_SDA, CUR_SEN_I2C_SCL,0x00, 0.00075, 60.95};
+
+    INA_226 safety_sensors[NUM_SAFETY_SENSORS] = {
+        INA_226 elbow_current_sensor(elbow_config), //example for elbow sensor, may initialize for each component
```
remove outdated comment

qinyang-bao commented on apps/safety/main.cpp:
```diff
@@ -84,13 +108,18 @@ void initCAN()
-        case EXPECTED_CAN_ID:
+        case SET_CONFIGURATIONS:
             pc.printf("Updating INA226 config\r\n");
+            //do stuff to update INA config register
```
A good habit is to write "TODO: " in front of things that you plan to do

qinyang-bao commented on lib/sensors/INA_226.cpp:
```diff
@@ -1,79 +1,159 @@
+    m_shunt_resistance = component_config.shunt_resistance;
+    m_sensor_address = component_config.sensor_address << 1; //7 bit address
+
+    i2c_(component_config.SDA_pinname, component_config.SCL_pinname);
```
naming of this member is not consistent with the style, should be m_i2c

qinyang-bao commented on lib/sensors/INA_226.cpp:
```diff
@@ -1,79 +1,159 @@
 
-    i2c_(SDA_pinname, SCK_pinname);
-    i2c_.frequency(i2c_frequency);
+    m_config_register = 0x00;
```
not sure why you decide to set these in the constructor instead of leaving them as constants?

FK3wong replied:
What would be the best way to do this? Initialize them inside the class using Static variable or just take them outside the class in the .cpp file?

qinyang-bao replied:
Preferably, values like register address should be constants as they don't change, and constants should be placed in the header file

ChristopherA057 commented on apps/safety/main.cpp:
```diff
@@ -11,97 +11,134 @@
+    ComponentConfig wrist_config = {CUR_SEN_I2C_SDA, CUR_SEN_I2C_SCL,0x00, 0.01, 8.05};
+    ComponentConfig bicep_config = {CUR_SEN_I2C_SDA, CUR_SEN_I2C_SCL,0x00, 0.015, 5.75};
+    ComponentConfig allen_key_config = {CUR_SEN_I2C_SDA, CUR_SEN_I2C_SCL,0x00, 0.022, 3.45};
+    ComponentConfig turntable_config = {CUR_SEN_I2C_SDA, CUR_SEN_I2C_SCL,0x00, 0.00075, 60.95};
```
Where are all these numbers from? Might be more readable if we have constexprs for them.

ChristopherA057 commented on lib/sensors/INA_226.cpp:
```diff
@@ -1,79 +1,149 @@
-    max_expected_current_ = max_expected_current;
-    sensor_address_ = sensor_address << 1;
+    m_max_expected_current = component_config.max_expected_current;
+    m_current_lsb = m_max_expected_current / pow(2,15);
```
Maybe you could replace the pow() command with the actual number and just explain it in a comment?

ChristopherA057 commented on lib/sensors/INA_226.cpp:
```diff
@@ -83,6 +83,19 @@ int INA_226::configureSensor(SensorModes configuration_bits)
+u_int16_t INA_226::readConfigRegister()
+{
+    char cmd[2] = {CONFIG_REGISTER, 0x00};
+    u_int16_t registerData = 0x00;
```
Tiny nitpick, but does this variable actually need to be instantiated before you set it on line 94?

FK3wong commented:
Waiting on Safety Test board to come in before writing more code on the current sensor. The safety APP may be deprecated due to the possible removal of a safety board.

FK3wong commented:
@wmmc88 I'm not quite sure what to do to fix the clang build issues. Could you lend me a hand and maybe suggest some things I could try to do to fix it?

wmmc88 commented:
> @wmmc88 I'm not quite sure what to do to fix the clang build issues. Could you lend me a hand and maybe suggest some things I could try to do to fix it?

You need to fix the formatting. either do it manually by looking at what failed in the log or use clangformat to do it for you. Either install clang-format 9 from the command line and run it on the files that fail or download a clangformat plugin//extension for your ide and run it.

jetkov commented:
@FK3wong Can you confirm EE status of safety functionality and if this PR should be dropped / code moved to another app project?

FK3wong commented:
On Pause as Testing still needs to be completed and Arm tasks are a priority
